sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+
+  byte *Ptr = nullptr;
+};
+
+} // namespace sycl
+```
+
+Here [`__builtin_unique_stable_name`][builtin-unique-stable-name]
+is a compiler built-in used to translate types to unique strings, which are
+used as symbolic IDs of specialization constants.
+
+[builtin-unique-stable-name]: https://github.com/intel/llvm/blob/sycl/clang/docs/LanguageExtensions.rst#__builtin_unique_stable_name
+
+`__sycl_getScalar2020SpecConstantValue<T>` and
+`__sycl_getComposite2020SpecConstant<T>` are functions with special names - they
+are declared in the headers but never defined. Calls to them are recognized by
+a special LLVM pass later and this is aforementioned special markup required for
+the compiler.
+Those functions accept three parameters:
+1. Symbolic ID of a specialization constant. Even though at SPIR-V level
+   specialization constants are identified by numeric IDs, we can't use them
+   here, because:
+   - Those IDs can't be generated by runtime, because they need to be encoded
+     into resulting SPIR-V device image
+   - Those IDs can't be generated by front-end compiler, because it only sees a
+     single translation unit at a time and therefore it can't assign unique IDs
+     to specialization constants from different translation units.
+
+   Therefore, the decision was made to use symbolic IDs as interface between the
+   compiler and runtime to connect SYCL identifiers of specialization constants
+   with SPIR-V identifiers of specialization constants.
+
+2. Default value of the specialization constant.
+   It is expected that at LLVM IR level the argument will contain a pointer to
+   a global variable with the initializer, which should be used as the default
+   value of the specialization constants.
+
+3. Pointer to a buffer, which will be used if native specialization constants
+   are not available. This pointer is described later in the section
+   corresponding to emulation of specialization constants.
+
+Compilation and subsequent linkage of the device code results in a number of
+`__sycl_getScalar2020SpecConstantValue` and
+`__sycl_getComposite2020SpecConstantValue` calls. Before generating a device
+binary, each linked device code LLVM IR module undergoes processing by
+`sycl-post-link` tool which can run LLVM IR passes before passing the module
+onto the SPIR-V translator.
+
+#### DPC++ Compiler: sycl-post-link tool
+
+As it is stated above, the only place where we can properly handle
+specialization constants is somewhere during or after linking device code from
+different translation units, so it happens in `sycl-post-link` tool.
+
+There is a `SpecConstantsPass` LLVM IR pass which:
+1. Assigns numeric IDs to specialization constants found in the linked module.
+2. Brings IR to the form expected by the SPIR-V translator.
+3. Collects and provides \<Symbolic ID\> =\> \<numeric IDs + additional info\>
+   mapping, which is later being used by DPC++ RT to set specialization constant
+   values provided by user.
```

Comment:
Where is this mapping information communicated?

Reply:
I think @romanovvlad describes it further below in the document.  Probably a forward reference, or saying that will be descibed in section ... would be sufficient here.

Reply:
It is communicated through the device binary properties mechanism, see "Collecting spec constants info and communicating it to DPC++ RT" below.

Looks like with SYCL2020 this can be [avoided](https://github.com/intel/llvm/pull/3331/files#r595187318).

Reply:
This section needs to be rewritten once we settle down on https://github.com/intel/llvm/pull/3331/files#r595187318

Reply:
Can we mark this comment as resolved and continue the discussion in https://github.com/intel/llvm/pull/3331#discussion_r595561411 ?